### ISSUE TITLE
weapondelayaction variable (forum poll)

### ISSUE THREAD

**Issue by JeffBobbo** [red-eclipse/base#198]
Move the interval after firing for switching and picking up weapons to this new variable from <weapon>delayattack.

http://redeclipse.net/forum/viewtopic.php?f=4&t=696

At the time of creating this, the poll has been running for a bit over three weeks (I forgot to set a closing date).

Votes were:
5 for <weapon>actiondelay
6 for weaponactiondelay
4 to leave it as it is

**Comment by shirepirate** [comment]
the problem with this is each of the actions you've mentioned are very different parts of balance.
the variables could probably be configured differently though, mostly it's incredibly annoying for delayattack to postpone item pickups.

to quote @qreeves :
> Instead of using items littered throughout a level, RE uses a series of complicated timers to balance the game out more evenly
> Attack delay is the time imposed on you while you are taking the force of the firing action, and in cases like the plasma and rifle it is essential to the way the balancing on those weapons work.
> You can't do any weapon related thing during this state.

while i do agree with all of this, I don't exactly see why delayattack prohibits switching weapons or picking up drops. delayattack only needs to prohibit the actual firing of **any** weapon until the delay ends, i don't think there needs to be any sort of timer set on switching weapons or prohibiting picking up drops.

considering reload also has it's own variable, it probably doesn't even need to prohibit this weapon action either. and while this would require a little math to adjust each weapon reload appropriately, i think it would make sense to have an actual separate delay for attack and reload which run simultaneously, rather than working in conjunction. i also suggest it be renamed to reloadtime.

so my opinion is to make these changes:

**/weapondelayattack** (will now affect *only* when *any* weapon may be fired next, instead of covering other actions)
**weaponreloadtime** (replaces delayreload, not hindered by any delay, but adjusted to an appropriate time for each weapon, basically add current delayattack + delayreload)
**no delays at all for switching between weapons or picking up item drops** (but remember, no abuse is possible because of delayattack)

these changes would probably break protocol anyway.

**Comment by acerspyro** [comment]
Even if they do break the protocol, it would finally force people who are still on 1.5.0 to update to 1.5.2. It would force them to figure out how to update.

**Comment by JeffBobbo** [comment]
@shirepirate I'm pretty sure the delays in place (weaponswitchdelay) to allow time for the animation to play out (as well as any possible balancing purposes), which is why there's a delay for picking up items. It'd look weird if you cut someone with sword2 while picking up a grenade. That being said the purpose of the topic was to change the current nature that the high RoF weapons get an advantage because of the way the system works.

Reload should stay as it is, otherwise you're just penalizing people for using weapons with long reloads and delayattack even more (rifle reload time would go up to 2350ms or 2500ms (r1/r2), zapper would be 2700ms for zapper2).

**Comment by shirepirate** [comment]
it's nice that the animation has a chance to play out, but this delay is too long. perhaps the animation could be sped up 2x and weaponswitchdelay changed to 200ms

the annoyance of nothing happening when you walk by a pickup and press E is also more important in my opinion than convincing animations. i don't think a delay is needed for pickups at all.

**Comment by qreeves** [comment]
This is much like the zoom issue. There is a HUGE difference between an idea that seems simple in your head compared to the complexity of the code that idea requires.

In this case, again, it comes down to the fact that the game tracks ALL weapon interactions based on `gamestate::weapstate[x]` (the last action performed), `gamestate::weaplast[x]` (the last time the action was performed), and `gamestate::weapwait[x]` (the delay length for the action).

Introducing any level of complexity to this system results in thousands of lines of code needing to be (re)written because it simply doesn't support it. I'm also not in favour of messing with the balance in this fashion as you'd need to review every weapon again for it.

Also, I should point out most weapon actions allow you to break out of the reload state (at the cost of having to start all over again when you reload next). The weapons used to be balanced more toward having higher reload delays than attack delays. I'm not sure why we've pushed the attack delays so high on some of the weapons, but really the simpler solution is to fix that.

**Comment by JeffBobbo** [comment]
If it's complex, we could just ignore the <weapon>delayattack timer as shire suggested for picking up items. Balance wise, this change has very little difference, we still have weaponswitchdelay and we could keep weapon switching as it is and just change item pickups. The issue I'm trying to address is what I pointed out in the thread, smg/flamer/zapper1 users get an advantage purely because their weapons have a quick rate of fire for actions that doesn't involve the weapon itself (such as picking up an item).

As for pushing for lesser delayattacks and larger delayreloads, I don't think this is a good idea. Play styles of players evolve over time as new tricks are learnt and new players with their own (new) play styles come to the game. What we've currently got is a large number of players who're good at using rifle. Reducing the delayreload would most likely result in pushes for nerfing the rifle (and similar weapons, like sg2) in other ways, such as damage reduction or increase spread which'll just devalue the weapon.

**Comment by shirepirate** [comment]
so, separating item pickups from how they are handled now as a weapon action is quite a lot of work, got it. not to be arrogant, but i'm still for eventually making this change. it's not handled very well now, bobbo brings a good point about smaller delays having an advantage (and longer reloads isn't a good way to fix this, it'd only make all the weapons more frustrating to use).

what about the separate issue of lowering weaponswitchdelay some? or is this affected by both weaponswitchdelay as well as weapon attack delays (that seems confusing)?

**Comment by qreeves** [comment]
`weaponswitchdelay` is just the delay that is imposed after you switch to a weapon, which is a special state that can override itself (`W_S_SWITCH`). Basically, you can keep switching weapons as fast as you want, but when you rest on your final selection you must wait out that delay before firing/reloading/etc. It is also used as the delay for after you pick up a weapon (`W_S_USE`), but that is a separate state of its own that can't be overridden.

The last special state case is reloading (`W_S_RELOAD`), which is also the most recent implementation. You all know this as the ability to break out of a reload cycle, and before this was added, you had to wait it out no matter what. Almost any action can override the reload state, the penalty being that you have to start all over again when you decide to reload. In terms of special cases, this is an easy state to track because it just resets the ammo to what it was when you break out of it.

The problem with the firing states (`W_S_PRIMARY`, `W_S_SECONDARY`) is that they are entirely timer dependant. There's nothing that can be reverted to when you break out of it. These states are supposed to only be the delay for which it takes to fire the weapon, and recover from its after effects (like what would be known as recoil, etc). They're simply not meant to be extremely long delays, they've gotten this way by extreme and random weapon balance attempts.

Now, the entire codebase assumes one thing: *a weapon that is not in use is considered idle* (`W_S_IDLE`). This gets automatically set and/or assumed by so many levels of the code, it isn't a trivial thing to start treating the state system as something that persists when not in use.

I'm not arguing that the delay can be a nuisance for picking up items, etc. I'm saying that overcomplicating the state system is a bad idea, from experience. I have tried to solve too many simple problems with grand code delusions, and they have always blown up in our faces.

So, what else can we do?

**Comment by JeffBobbo** [comment]
If that's the case, I don't think there's anything we really can do without fixing this state system.

Would it be much work to expand the state system to store states for all the weapons individually?

**Comment by freezurbern** [comment]
>Would it be much work to expand the state system to store states for all the weapons individually?

Yay, more data to send.

**Comment by shirepirate** [comment]
>> Basically, you can keep switching weapons as fast as you want, but when you rest on your final selection you must wait out that delay before firing/reloading/etc.

this doesn't seem to be the case. if i recall switching between weapons could be done lot quicker in 1.4. or are you saying that it's feasible, but weaponswitchdelay happens to be set at 500ms?

**Comment by qreeves** [comment]
> switching between weapons could be done lot quicker in 1.4. or are you saying that it's feasible, but weaponswitchdelay happens to be set at 500ms?

Yeah, that is actually a client side variable (`weapselectdelay`) to stop the weapons switching *too* fast, especially when using the scroll wheel. It defaults to 250ms and is purely client controlled.

**Comment by shirepirate** [comment]
ok, just lowering that to 200ms seems to help a lot actually.

i'm out of ideas for the pickups, basically the way it would work ideally requires far too much coding for little gain that's better spent on other things.

**Comment by JeffBobbo** [comment]
So I've been looking at the code and I'm pretty sure this is doable without much change at all.

We already store the states for all the weapons individually, so we can adapt clientstate::weapswitch() to keep the state of the old weapon (and if you wanted, you could use it to propagate the delayattack to the new weapon). Weapon switches between weapons you already have could stay as they are.
